### DISCUSSION TITLE
\qty 为什么不可以替换 \SI

### DISCUSSION THREAD

**Discussion by earthmoon** [whutug/whu-thesis#113] [Q&A]
你好，我在模板中用\SI是正常的，将\SI 改为\qty 就报错，为什么呢？

**Comment by zepinglee** [comment]
`\qty` 是 `siunitx` v3.0 的用法，`\SI` 是 v2.x 的用法。

**Reply by earthmoon** [reply]
我们的模板不支持3.0么

**Reply by zepinglee** [reply]
跟模板没有太大关系，你需要检查一下本地的 `siunitx` 的版本。另外“将\SI 改为\qty 就报错”最好改为给出具体的报错信息，以方便讨论。

**Reply by earthmoon** [reply]
版本挺高的：unitx/siunitx.sty  Package: siunitx 2022-02-15 v3.0.49 A comprehensive (SI) units package

> [{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Missing $ inserted.\n<inserted text> \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Extra }, or forgotten $.\n\\@quantity code ...8}{}{|#8|}} \\ifnum \\z@ =`{\\fi }\n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Undefined control sequence.\n\\meter ->\\ERROR \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Undefined control sequence.\n\\cubed ->\\ERROR \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Undefined control sequence.\n\\per ->\\ERROR \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Undefined control sequence.\n\\second ->\\ERROR \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "You can't use `\\spacefactor' in math mode.\n\\blx@leavevmode ...\\else \\leavevmode \\spacefactor \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "You can't use `\\spacefactor' in math mode.\n\\blx@leavevmode ...\\else \\leavevmode \\spacefactor \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "You can't use `\\spacefactor' in math mode.\n\\blx@leavevmode ...\\else \\leavevmode \\spacefactor \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "You can't use `\\spacefactor' in math mode.\n\\blx@leavevmode ...\\else \\leavevmode \\spacefactor \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 14,
> 	"startColumn": 1,
> 	"endLineNumber": 14,
> 	"endColumn": 65536
> },{
> 	"resource": "/g:/OneDrive - Buddhist Tai Hung College/0.1.Studying/Ph.D. Thesis/whu-thesis-0.6d/pages/chapter2.tex",
> 	"owner": "LaTeX",
> 	"severity": 8,
> 	"message": "Missing $ inserted.\n<inserted text> \n",
> 	"source": "LaTeX",
> 	"startLineNumber": 15,
> 	"startColumn": 1,
> 	"endLineNumber": 15,
> 	"endColumn": 65536
> }]
谢谢了！

**Reply by sikouhjw** [reply]
任何提问请给出最小工作示例和命令行编译的报错。

**Reply by earthmoon** [reply]
> 任何提问请给出最小工作示例和命令行编译的报错。

您好，tex文档（我修改过的）在这里：
[全部文件](https://bthc-my.sharepoint.com/:f:/g/personal/mr_iem_im/EkcqvqgyPAVOgCTIA7GHAxsB6Act1fZG5ndU0NdhuNmQqQ?e=m0aMgr)

**Comment by tanukihee** [comment]
因为用了 physics

si 和 physics 一起用的时候有一条警告，我特意留着了，你没看到吗

**Reply by earthmoon** [reply]
> 因为用了 physics
> 
> si 和 physics 一起用的时候有一条警告，我特意留着了，你没看到吗

这个文档的警告有时有几十条，警告都要处理么。那估计要挺大工作量了。physics不能和qty一起用吗

**Reply by tanukihee** [reply]
不要你处理，你至少要看一看吧

physics 已经定义了 `\qty`，si 检测到 physics 包加载时，就不会再重复定义 `\qty` 了，而是使用旧版命令 `\SI`。

遇到问题之前先看看文档吧

**Comment by earthmoon** [comment]
好的吧，看来qrt不能用了，那就si凑合吧。
________________________________

**Reply by tanukihee** [reply]
不是 qty 不能用，是有其他的用途了

**Comment by earthmoon** [comment]
嗯嗯，这个有替代品也还好～
________________________________

**Comment by zepinglee** [comment]
我不建议在模板中默认加载 `siunitx` 后者 `physics`，比如数学系的论文就不太可能用到这两个包的功能。让用户在导言区自行调用并负责的方式对模板而言更为简单。

**Reply by tanukihee** [reply]
的确如此，本来这两个包蛮好的，结果 si 来个大更新……当时我的论文都写完了也就懒得管了，后面的屎山请 @xkwxdyy 佬多辛苦一下了（逃

**Reply by zepinglee** [reply]
我倾向于这样的方式：模板在默认情况下只加载配置格式所必需的包（如 `geometry`, `fancyhdr`, `titletoc`），额外的包需要用户在导言区手动加载，但模板会通过 `filehook` 的方式在用户加载宏包后进行相应的设置以适配模板的格式，比如（`longtable`, `siunitx`, `hyperref` 等）。这样也能给予用户更多的灵活性，尤其是那些有多个可选宏包的功能方面，比如参考文献的 `natbib` 和 `biblatex`，数学定理的 `amsthm` 和 `ntheorem`，算法的 `algorithms` 和 `algorithm2e`。

**Comment by earthmoon** [comment]
SI它更新的时候不考虑这些吗，那冲突会越来越多吧

________________________________

**Reply by tanukihee** [reply]
那你得去问 si 的作者为什么搞这种 breaking change 了

至少是从 2013 年开始筹划的 https://github.com/josephwright/siunitx/issues/116
